akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -548,3 +545,228 @@ final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val completion = Promise[Done]()
+
+    val logic: GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var resource: S = _
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        try {
+          resource = open()
+        } catch {
+          case NonFatal(e) ⇒
+            completion.failure(e)
+            throw e
+        }
+        pull(in)
+      }
+
+      final override def onPush(): Unit = {
+        try {
+          writeData(resource, grab(in))
+          pull(in)
+        } catch {
+          case NonFatal(ex) ⇒ decider(ex) match {
+            case Supervision.Stop ⇒
+              close(resource)
+              doFailStage(ex)
+            case Supervision.Restart ⇒
+              restartState()
+              pull(in)
+            case Supervision.Resume ⇒
+              pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = closeStage()
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        close(resource)
+        doFailStage(ex)
+      }
+
+      private def restartState(): Unit = {
+        close(resource)
+        resource = open()
+      }
+
+      private def closeStage(): Unit =
+        try {
+          close(resource)
+          doCompleteStage()
+        } catch {
+          case NonFatal(ex) ⇒ doFailStage(ex)
+        }
+
+      private def doFailStage(th: Throwable): Unit = {
+        completion.failure(th)
+        failStage(th)
+      }
+
+      private def doCompleteStage(): Unit = {
+        completion.success(Done)
+        completeStage()
+      }
+    }
+
+    (logic, completion.future)
+  }
+
+  override def toString = "FoldResourceSink"
+}
+
+/**
+ * INTERNAL API
+ */
+final class FoldResourceSinkAsync[T, S](
+  open:      () ⇒ Future[S],
+  writeData: (S, T) ⇒ Future[Unit],
+  close:     (S) ⇒ Future[Unit])
+  extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.out")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val completion = Promise[Done]()
+    val logic: GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var resource = Promise[S]()
+      var writeFuture: Future[Unit] = Future.successful(())
+
+      implicit val context = ExecutionContexts.sameThreadExecutionContext
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        createStream()
+        setKeepGoing(true)
```

Comment:
I think it would be better to create both the async callbacks in preStart, and keep a reference to their invoke-method, that way you avoid extra allocations every time you want to use the async callback.